src/core/thread/address_resolver.cpp

```diff
@@ -170,17 +170,64 @@ AddressResolver::Cache *AddressResolver::NewCacheEntry(void)
     return rval;
 }
 
-void AddressResolver::MarkCacheEntryAsUsed(Cache &aEntry)
+void AddressResolver::MoveCacheEntryAt(Cache &aEntry, uint8_t aAge)
 {
-    for (int i = 0; i < kCacheEntries; i++)
+    VerifyOrExit(aEntry.mAge != aAge);
+
+    if (aEntry.mAge > aAge)
+    {
+        for (int i = 0; i < kCacheEntries; i++)
+        {
+            if (mCache[i].mAge >= aAge && mCache[i].mAge < aEntry.mAge)
+            {
+                mCache[i].mAge++;
+            }
+        }
+    }
+    else
+    {
+        for (int i = 0; i < kCacheEntries; i++)
+        {
+            if (mCache[i].mAge > aEntry.mAge && mCache[i].mAge <= aAge)
+            {
+                mCache[i].mAge--;
+            }
+        }
+    }
+
+    aEntry.mAge = aAge;
+
+exit:
+    return;
+}
+
+void AddressResolver::MoveCacheEntryBehindCached(Cache &aEntry)
+{
+    uint8_t i;
+    uint8_t lastInCachedState = 0;
+
+    for (i = 0; i < kCacheEntries; i++)
     {
-        if (mCache[i].mAge < aEntry.mAge)
+        if (mCache[i].mState != Cache::kStateCached || mCache[i].mAge == aEntry.mAge)
+        {
+            continue;
+        }
+
+        if (mCache[i].mAge > lastInCachedState)
         {
-            mCache[i].mAge++;
+            lastInCachedState = mCache[i].mAge;
         }
     }
 
-    aEntry.mAge = 0;
+    if (i == kCacheEntries)
```

Comment:
There is no `break` in the above `for` loop, so when we come out of it, wouldn't `i` always be `kCacheEntries`?

I think we may need to track if we found any entry in cached state or not (with a bool var maybe?) so we can know if `lastInCachedState was set or not?

Reply:
Thanks, heh, before filing this PR I actually changed implementation from your proposed approach to reduce one variable.. apparently, it's not always good to optimize everything :)